### PR TITLE
cuda: accelerate ConvRot with cooperative decode

### PR DESCRIPTION
**Pull request by mgeldi**
## Summary

- replace the one-block-per-row CR decode fallback with a cooperative inverse-ConvRot matvec for Q5_CR, Q6_CR, and Q8_CR tiny-output cases while retaining the Pascal reference path
- fuse the 256-value ConvRot activation transform into Q8_1 preparation for the existing MMVQ/MMQ kernels, avoiding the intermediate FP32 rotated-activation buffer
- add whole-graph dependency-ordering, adversarial-input, producer-boundary, tiny-shape, production-shape, and performance coverage

## Context

This follows #312 and implements the cooperative inverse-transform/dot-product direction discussed after the correctness fallback recovered coherent output but left decode substantially behind the ordinary quantized path.

The standard path keeps CR weights in their compatible Q8_0, Q5_0, or Q6_K physical layouts and rotates plus quantizes activations directly for MMVQ/MMQ. Unsupported consumers retain a lazy FP32 fallback. Small output matrices use the exact cooperative inverse-weight path.

The corruption mechanism encountered during development was PDL predecessor ordering combined with raw `__restrict__` aliases. The new kernels use the backend's PDL-safe pointer contract and include a `scale -> CR MUL_MAT` graph regression.

## Verification

Tested from base `1c2034f4a2518ca89ba3f42eee7df6830f8b4bb2` on an RTX 5090 with CUDA 13.3:

- full CUDA0 backend sweep: 18,390/18,390 passed
- ConvRot CUDA operator suite: 204/204 with CUDA graphs enabled and 204/204 in a clean graphs-disabled build
- Compute Sanitizer memcheck and synccheck: all 204 ConvRot cases passed with zero reported errors
- host Q8_CR, Q5_CR, and Q6_CR quantization coverage passed
- standalone TurboQuant accuracy and chunked-dequantization suite passed
- portable CUDA compilation covered 75-virtual, 80-virtual, 86-real, 89-real, 90-virtual, 120a-real, and 121a-real
- two deterministic 64-token generations produced identical normalized token streams
- exact inverse-reference comparison: 0.9999577 final-logit cosine similarity, 0.965% relative L2, identical argmax, and 20/20 top-token overlap

## Performance

Counterbalanced seven-repetition Qwen3.8-27B Q6_CR/Q6_K runs with `p=2048`, `n=512`, `b=ub=256`, F16 KV, flash attention, and 99 GPU layers retained a geometric mean of:

- 107.60% prompt throughput
- 98.02% decode throughput

These ratios are workload-specific. Matched operator-only throughput retention (`n=1` / `n=256`) was Q8_CR 95.83% / 94.82%, Q5_CR 93.56% / 96.90%, and Q6_CR 96.29% / 96.91%.

Runtime testing was limited to Blackwell/CUDA 13.3. CUDA 12, pre-Blackwell runtime hardware, HIP, MUSA, and full-model Q5_CR/Q8_CR artifacts were not available.

## AI assistance

AI materially assisted the design analysis, implementation, debugging, tests, benchmarks, review, and documentation. I manually reviewed every changed line, understand the ConvRot architecture and submitted implementation, verified the claims above, and am prepared to explain and maintain the changes.

### CODE REVIEW

**Comment by giveen** [comment]
@mgeldi as soon as my GPU is free I'll test this out, thank you so much for this follow up to fix an issue I couldnt figure out. I am hoping this restores the decode speed.

**Comment by giveen** [comment]
@TheTom  @mgeldi 

## Decode speed (llama-bench)

Command: ./build/bin/llama-bench -m <model> -p 2048 -n 512 -b 256 -ub 256
         -ctk f16 -ctv f16 -fa on -ngl 99 -r 1

| Model  | size     | pp2048    | tg512 decode | Decode vs Q8_0 |
|--------|----------|-----------|--------------|----------------|
| Q8_0   | 26.62 GiB| 3127.33   | 51.81        | 100%           |
| Q8_CR  | 26.62 GiB| 2957.85   | 50.91        | 98.3%          |
| Q6_CR  | 20.88 GiB| 2457.45   | 58.72        | 113%           |
| Q5_CR  | 17.50 GiB| 3029.58   | 70.47        | 136%           |

Headline: decode went from 9.13 t/s (pre-PR fallback, reported in pr312 report) to
50-70 t/s - at or above the Q8_0 reference. The CR decode slowdown is fixed.
Note: Q6_CR/Q5_CR are smaller files (20.9/17.5 GiB vs 26.6 GiB), so part of the gain
is bandwidth, but the cooperative kernel is no longer the bottleneck.

## KLD / PPL (vs Q8_0 base)

Base logits file generated fresh: /mnt/storage/models/qwen3.8/q8_0_logits.kld (73.4 GiB)
Command: 

```
./build/bin/llama-perplexity -m <model> -f wiki.test.raw -c 512 -b 2048 -t 4
         -ngl 99 --kl-divergence --kl-divergence-base q8_0_logits.kld
580 chunks, n_ctx=512, batch_size=2048, n_seq=4, wikitext-2 test set
```

| Model  | Mean KLD      | PPL(Q)    | PPL(Q)/PPL(base) | Same top p | RMS dp |
|--------|---------------|-----------|------------------|------------|--------|
| Q8_CR  | 0.001327      | 6.957961  | 1.000538         | 98.487%    | 1.048% |
| Q6_CR  | 0.003221      | 6.967321  | 1.001884         | 97.454%    | 1.569% |
| Q5_CR  | 0.011753      | 6.994244  | 1.005756         | 94.907%    | 3.024% |

Q8_CR and Q6_CR are excellent (KLD < 0.01). Q5_CR (3-bit) is coherent with a slightly
higher but acceptable KLD. The activation-rotation fusion did not regress output quality.


I also did a real world test
```bash
build/bin/llama-server   -m /mnt/storage/models/qwen3.8/Qwen3.8-27B-Q5_CR.gguf   --mmproj /mnt/storage/models/qwen3.8/mmproj-Qwen3.8-27B-Q8.gguf   --no-mmproj-offload -md /mnt/storage/models/qwen3.8/MTP/mtp-Qwen3.8-27B-Q4_0.gguf  -c 226000   -ctk q8_0   -ctv turbo4   -ngl all   -fa on   --parallel 1   --load-mode mmap+mlock   --host 0.0.0.0   --port 8080   --spec-type draft-mtp   --spec-chain 8   --spec-draft-p-min 0   --reasoning on   --reasoning-preserve   --temp 1.0   --top-k 20   --top-p 0.95   --min-p 0   --chat-template-kwargs '{"reasoning_effort":"medium"}'   --chat-template-file /mnt/storage/llama-server/qwen3.6-chat-template/chat_template.jinja --alias "Qwen3.8-Q5"
```
**PROMPT**
```write a snake game in python and pygame```

Web browser UI
Coherent response and code output

```bash
0.34.101.330 I slot print_timing: id  0 | task 26 | prompt eval time =     253.04 ms /    21 tokens (   12.05 ms per token,    82.99 tokens per second)
0.34.101.332 I slot print_timing: id  0 | task 26 |        eval time =   15135.82 ms /  1978 tokens (    7.65 ms per token,   130.68 tokens per second)
0.34.101.333 I slot print_timing: id  0 | task 26 |       total time =   15388.86 ms /  1999 tokens
0.34.101.334 I slot print_timing: id  0 | task 26 |    graphs reused =        537
0.34.101.338 I slot print_timing: id  0 | task 26 | draft acceptance = 0.34751 ( 1454 accepted /  4184 generated), mean len =  3.78
0.34.101.365 I slot      release: id  0 | task 26 | stop processing: n_tokens = 2105, truncated = 0
```

**Comment by TheTom** [comment]
Independent Spark validation is clean on NVIDIA GB10 with CUDA 13.0: the PR builds for sm_121a and the complete ConvRot CUDA operator suite passes 204/204 with graphs enabled. That includes Q5_CR, Q6_CR, Q8_CR, adversarial inputs, producer boundaries, tiny shapes, and production shapes.

Giveen’s 5090 full-model results close the remaining gap: 9.13 tok/s is back to 50.91, 58.72, and 70.47 tok/s, with the KLD/PPL improvements intact and a coherent 226k MTP run at 130.68 tok/s. This clears the #312 performance blocker. Merging.

**Review by TheTom — APPROVED** [review]
Reviewed and independently validated on the Spark. 204/204 ConvRot CUDA tests pass.

**Comment by TheTom** [comment]
Post-merge integration check on bd9bd1b47 is clean: 204/204 ConvRot tests and 76/76 TQ MoE tests pass together on the Spark. Thanks @mgeldi and @giveen. This is the result we needed.

**Comment by TheTom** [comment]
@mgeldi Thank you for taking the correctness fallback all the way to a production-speed implementation, including the adversarial coverage and performance work. @giveen thank you for doing the 5090 A/B, full KLD/PPL run, and real MTP workload. That combination gave us enough evidence to merge confidently.
